### PR TITLE
Feature/frans lukas/2319 search to fragment

### LINKED ISSUE #2319: Convert SearchActivity to fragment
**Is your feature request related to a problem? Please describe.**

#2235

### PR DESCRIPTION
**Pull request by Frans-Lukas**
Fixes #2319

So much had to change to make this work with navigation. It will also need a bit of a re-write once tab restoration is properly fixed in #2288 .

### CODE REVIEW

**Comment by Frans-Lukas** [comment]
SearchViewModel tests needs a look.

**Comment by s-ayush2903** [comment]
Hey @Frans-Lukas you've converted almost every activity to fragment, please leave some work for me too(like currently there is `ErrorActivity` left probably), i've been trying implementing the conversion from activity to fragment locally on my machine(to get some clear understanding of codebase) and you've already done a ton of work!

**Comment by Frans-Lukas** [comment]
Hey, yes it is both fortunate and unfortunate that I've done all of these conversion tickets. I have done them because these tickets are clear and takes somewhat long to do, and I need things to do because of GSoC. ErrorActivity will not be converted to a fragment, see #2271 . 

However, if you want, I am having issues with #2281 (localfiletransfer to fragment). If you can wait until #2254 (java -> kotlin) is merged, I think you can take over #2281 ticket.

**Comment by Frans-Lukas** [comment]
Also, CustomDownload -> fragment is without a ticket as @macgills mentions in #2235

**Comment by s-ayush2903** [comment]
Cool, thanks i'll be taking it up as soon as i get some more tests running for the module

**Comment by Frans-Lukas** [comment]
Another issue that search no longer differs from openedFromTab and not openedFromTab. (reminder for myself)

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SearchInPreviousScreen.kt
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SearchInPreviousScreen.kt:29]: https://stackoverflow.com/a/60757744/12225251 not sure if this is much of an option?

Comment (Frans-Lukas) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SearchInPreviousScreen.kt:29]: hmm, will check out.

Comment (Frans-Lukas) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SearchInPreviousScreen.kt]: It works very nicely with one exception. The fragment result is not cleared on observation. This means that if the user first searches inside a page, and then opens a page, the search in page will open as well. I solved it by manually clearing the observable ID after observation, but I am suspecting this is not the correct way to do it. Thoughts? 

```
  fun <T> Activity.observeNavigationResult(
    key: String,
    owner: LifecycleOwner,
    observer: Observer<T>
  ) {
    getObservableNavigationResult<T>(key)?.observe(owner) {
      observer.onChanged(it)
      asCoreMainActivity().consumeObservable<T>(key) <------------------
    }
  }
  fun <T> Activity.consumeObservable(key: String = "result") =
    asCoreMainActivity().navController.currentBackStackEntry?.savedStateHandle?.remove<T>(key)

```

Comment (Frans-Lukas) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SearchInPreviousScreen.kt]:
```kotlin
    requireActivity().observeNavigationResult<SearchItemToOpen>(
      TAG_FILE_SEARCHED,
      viewLifecycleOwner,
      Observer(::openSearchItem)
    )
```

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SearchInPreviousScreen.kt]: You might need to present a more bug style report for me to understand this quickly

**Comment by macgills** [comment]
failing the unit tests

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SearchInPreviousScreen.kt
Comment (Frans-Lukas) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SearchInPreviousScreen.kt]: Good idea!

**Describe the bug**

Navigation results remain in `savedStateHandle`, the symptom of this is that on entry to readerFragment, the search bar will always open. 

**Expected behavior**

Once a navigation result has been read, it should be removed from the `savedStateHandle`. 

**Steps to reproduce the behavior:**
1. Enter search from the reader fragment.
2. Type "t" and click "Find in page".
3. Notice that the search bar opens in the reader fragment, and that you can navigate between found words. This is as expected.
4. Close the search bar by pressing the back button.
5. Open search again.
6. Search for a page, and open the page.
7. Notice that the page opens, but also the searchbar with the search phrase, "t" that was typed in step 2 is open. 


The above error is fixed by making sure the navigationResult is cleared from `savedStateHandle` after being read.

**Comment by Frans-Lukas** [comment]
I'll take a look at the inst. tests too..

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/reader/KiwixReaderFragment.kt
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/reader/KiwixReaderFragment.kt:103]: `apply` isn't quite idiomatic here, `let` is probably what we are looking for?

File: core/src/main/java/org/kiwix/kiwixmobile/core/extensions/ActivityExtensions.kt
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/extensions/ActivityExtensions.kt:107]: maybe this can be a property `coreMainActivity` instead

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java:1225]: a ternary would look nicer here maybe

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java:1241]: ternary

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java:1587]: some extra space here

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/MainMenu.kt
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/MainMenu.kt:163]: `apply` isn't so idiomatic, in this case I think you can just use an `if` or replace the apply with a `let`

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/ZimReaderContainerUtils.kt
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/utils/ZimReaderContainerUtils.kt:35]: this if doesn't need brackets and might fit on 1 line

File: core/src/main/res/layout/activity_search.xml
Comment (macgills) [inline core/src/main/res/layout/activity_search.xml:13]: commented out code

File: core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/SearchViewModelTest.kt
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/SearchViewModelTest.kt:269]: this name needs updating

**Comment by macgills** [comment]
@Frans-Lukas were we going to support deep linking via the widget in this ticket? or is that a different ticket?

Instrumentation seems to have widespread failures

**Comment by Frans-Lukas** [comment]
Hmm, I remember you asking about the widget for this ticket. I'd prefer it to be another ticket to make this mergable quicker. 

Usually the activity -> fragments causes many inst. tests to fail cascading from a single failing test. We'll see.

File: core/src/main/java/org/kiwix/kiwixmobile/core/extensions/ActivityExtensions.kt
Comment (Frans-Lukas) [reply core/src/main/java/org/kiwix/kiwixmobile/core/extensions/ActivityExtensions.kt:107]: Clean suggestion!

**Comment by macgills** [comment]
@Frans-Lukas now it is the unit tests that fail

File: app/src/androidTest/java/org/kiwix/kiwixmobile/intro/IntroFragmentTest.kt
Comment (macgills) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/intro/IntroFragmentTest.kt:35]: we can probably delete this, we should probably even go back to ignoring this test because it is so troublesome but that is another ticket

File: core/src/main/java/org/kiwix/kiwixmobile/core/extensions/ActivityExtensions.kt
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/extensions/ActivityExtensions.kt:42]: `coreMainActivity` is an okay name I think

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java:1243]: this doesn't look so nice, try

```
condition
? trueVale
: falseValue
```

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java:1392]: We can still open bookmarks and history items, yeah?

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchFragment.kt
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchFragment.kt:107]: I am also not against an extension on Fragment `coreMainActivity` feels like we may repeat this line a bit

File: core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/SearchViewModelTest.kt
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/SearchViewModelTest.kt:269]: this is still `CreatedWithIntent`

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java
Comment (Frans-Lukas) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/CoreReaderFragment.java:1392]: Yes, this fragment does not use any activity for result and bookmarks and history is handled by navigation arguments.

**Comment by Frans-Lukas** [comment]
I am not getting these unit test errors. Never seen similar ones before, have you @macgills ? `java.lang.instrument.IllegalClassFormatException: Error while instrumenting org/kiwix/kiwixmobile/core/zim_manager/Language.` 

`Caused by: java.lang.IllegalStateException: Cannot process instrumented class org/kiwix/kiwixmobile/core/zim_manager/Language. Please supply original non-instrumented classes.` 

`    java.lang.instrument.IllegalClassFormatException: Error while instrumenting org/kiwix/kiwixmobile/core/extensions/ViewGroupExtensions.`

`    Caused by: java.lang.IllegalStateException: Cannot process instrumented class org/kiwix/kiwixmobile/core/extensions/ViewGroupExtensions. Please supply original non-instrumented classes.`

**Comment by macgills** [comment]
Those aren't the real error

```
SearchInPreviousScreenTest > invoke with returns positive result with string to previous screen() FAILED
    java.lang.AssertionError at SearchInPreviousScreenTest.kt:37
```

those lines are just an annoyance when gnerating coverage. I have a ticket about making our tasks upload proper artifacts but it hasn't been prioritised

**Comment by Frans-Lukas** [comment]
Great! Thanks. I am used to being able to easily identify what test is failing, apparently not now.

**Review by macgills — APPROVED** [review]

**Comment by macgills** [comment]
agh, I was too hasty in approving this. The keyboard is much more obnoxious now, doesn't go away when I leave search, I'll make a ticket

**Pull request merged**
